### ISSUE TITLE
Consider adding a generic "should_skip" method

### ISSUE THREAD

**Issue by erickt** [serde-rs/serde#213]
Our current mechanism to skip serializing a value is a little flaky. Right now we have `#[serde(should_skip_if_empty)]` and `#[serde(should_skip_if_none)]`, which when put on a struct field, generate a condition check (`value.is_empty()` or `value.is_none()`) that is performed before we serialize the field's name. This is a little ugly and unhygenic, because we're just assuming that there is some method named `is_empty` or `is_none`. We'd also require adding new attributes each time we want to add some other check.

One potentially more robust approach would be to add a method to the `Serialize` trait that would allow a type to opt out of serializing for some reason:

```rust
trait Serialize {
    ...
    fn should_serialize(&self) -> bool { true } // default to always serializing this value
    ...
}
```

One downside with this approach is that it'd be hard to know a priori if a generic type, like `HashMap`, should serialize itself if it is empty or not. So the other option is something like #208, where we create a trait that works with `#[serde(skip_serializing_if_empty)]` that triggers this conditional check.

I'm starting to lean more towards #208, but I wanted to make this ticket in case anyone else had more thoughts on the matter.

**Comment by oli-obk** [comment]
> One downside with this approach is that it'd be hard to know a priori if a generic type, like HashMap, should serialize itself if it is empty or not. 

I don't think it should be type-dependent. This should be field-dependent. Even if it were a use-case to have a type choose whether it should be serialized, then the issue is entirely orthogonal to skipping fields. Otherwise we end up needing a `serialize_even_if_empty` xD

**Comment by erickt** [comment]
This was implemented in #238!
